src/server/graceful.rs

```diff
@@ -0,0 +1,447 @@
+//! Utility to gracefully shutdown a server.
+//!
+//! This module provides a [`GracefulShutdown`] type,
+//! which can be used to gracefully shutdown a server.
+//!
+//! See <https://github.com/hyperium/hyper-util/blob/master/examples/server_graceful.rs>
+//! for an example of how to use this.
+
+use std::{
+    fmt::{self, Debug},
+    future::Future,
+    pin::Pin,
+    task::{self, Poll},
+};
+
+use pin_project_lite::pin_project;
+use tokio::sync::watch;
+
+/// A graceful shutdown utility
+pub struct GracefulShutdown {
+    tx: watch::Sender<()>,
+    rx: watch::Receiver<()>,
+}
+
+
+impl GracefulShutdown {
+    /// Create a new graceful shutdown helper.
+    pub fn new() -> Self {
+        let (tx, rx) = watch::channel(());
+        Self { tx, rx }
+    }
+
+    /// Wrap a future for graceful shutdown watching.
+    pub fn watch<C: GracefulConnection>(&self, conn: C) -> impl Future<Output=C::Output> {
```

Comment:
I'm actually thinking of making this `&mut self`, even though it's not _required_, but it would allow us to change internals.

Reply:
What would be a hypoteical practical case where you need &mut in a scenario like this?

Reply:
If we changed from using a channel to something that mutates `self` non-atomically. It is hard to think of how that could work...